enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,463 @@
+
+Operators managed by OLM would benefit from the ability to communicate their state directly to OLM through a supported channel. Operators could communicate complex conditions and influence how OLM manages the operator, dramatically improving user experience.
+
+A core requirement of this enhancement is to provide operators with the ability to communicate to OLM that the operator should not be upgraded while a critical process is being executed. Consider the following scenario based on OLM today:
+
+> An operator installed with automatic updates is executing a critical process that could place the cluster in a bad state if interrupted. While this critical process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the critical process to finish and places the cluster in an unrecoverable state.
+
+OLM can provide a significantly better user experience by creating a channel for the operator to communicate state that OLM cannot infer otherwise.
+
+### Goals
+
+- Provide operators with the means to explicitly communicate a condition that OLM cannot infer on its own.
+- Allow operators to define specific conditions which will influence OLM behavior, such as the "Upgrade Readiness" condition defined above.
+- Preserve existing behavior should the operator not communicate its upgradeable status.
+
+### Non-Goals
+
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate conditions to OLM that it cannot infer otherwise.
+
+#### Story 2
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade or uninstall my operator until I tell it otherwise.
+
+## Design Details
+
+Operators will communicate with OLM by setting [conditions](https://dev.to/maelvls/what-the-heck-are-kubernetes-conditions-for-4je7) on an owned CustomResource. Conditions are used by many Kubernetes Resources to provide users with additional insight into the state of the resource. For those unfamiliar with conditions, they typically follow the following format:
+
+| Field | Description |
+| ----- | ----------- |
+| Type |  A one-word description of the condition. |
+| Status | Status is the status of the condition. Can be True, False, Unknown. |
+| Reason | Unique, one-word, CamelCase reason for the condition's last transition. |
+| Message | Human-readable message indicating details about last transition. Typically used by cluster admins for debugging purposes. |
+| LastTransitionTime | Last time the condition transitioned from one status to another. |
```

Comment:
This (or a lastHeartbeatTime) might be a useful field to enforce a hard or soft grace-period. A NotInterruptible condition must be updated at least every 30 minutes, for example. If not:
1) Soft grace-period - let the user know through OLM that the operator needs their attention.
2) Hard grace-period - assume the process in unrecoverable and allow the upgrade to continue anyway.

Reply:
This is an interesting idea. My only concern is an operator that does not update the condition within the Hard grace-period could lead to severe consequences that OLM can't anticipate. If we decide to proceed down this route, it seems like an Operator Author should be able to opt-out of the hard grace-period. Thoughts?

Reply:
Allowing them to opt out is worth considering, but should probably be discouraged since it instills a false confidence. A workload being able to pick up where it left off without corrupting data is part of being a k8s workload. I see this enhancement as OLM being polite. When a node is drained, or a pod evicted, the operator will not receive the same politeness (just a normal k8s graceperiod before SIGKILL).

Reply:
> A workload being able to pick up where it left off without corrupting data is part of being a k8s workload. 

+1000, unfortunately we've seen time and again that some operators are unable to deliver their operator in this format.

> I see this enhancement as OLM being polite. When a node is drained, or a pod evicted, the operator will not receive the same politeness (just a normal k8s graceperiod before SIGKILL).

I'm not convinced that this is the right approach. If operators are truly successful and used by our customers to drive their business there could easily be a situation where OLM ignoring a critical condition of an operator causes direct revenue loss. Your suggestion might be the right approach, but I want other to share their opinions as well. CC @shawn-hurley @ecordell

Reply:
> IMO, it would be an admins choice to enable a hard grace-period or a soft grace-period and they would be responsible for the risks associated with the workloads they run. I could see lots of folks being very happy having a hard grace-period while others may very much want a soft grace period. The key here would be documenting the user choice.
I don't think that operator author would get the choice here, they get the ability to state some things and if the user want's to ignore or take other actions I don't think OLM should get in the way of that

100%

> If this is something we are debating, are we adding a user story for it or capturing it for a follow up somewhere? I don't want to lose the idea.

I can update this enhancement or create it as an epic we introduce later to keep the scope of this enhancement smaller - whichever you prefer.

Reply:
I think we could get something similar -- w/o putting the onus on the operator to supply the heartbeat -- by supporting "condition source" combinations; e.g. `upgradeable=f(CR, pod)`, where perhaps the pod being ready is required for `f(CR, pod)=true`.